environs/manual/winrmprovisioner/winrmprovisioner.go

```diff
@@ -0,0 +1,577 @@
+New-Item -Path WSMan:\localhost\ClientCertificate -Issuer $clientcert.Thumbprint -Subject $subject -Uri * -Credential $cred -Force
+`
+
+// gatherMachineParams collects all the information we know about the machine
+// we are about to provision. It will winrm into that machine as the provision user
+// The hostname supplied should not include a username.
+// If we can, we will reverse lookup the hostname by its IP address, and use
+// the DNS resolved name, rather than the name that was supplied
+func gatherMachineParams(hostname string, cli manual.WinrmClientAPI) (*params.AddMachineParams, error) {
+	// Generate a unique nonce for the machine.
+	uuid, err := utils.NewUUID()
+	if err != nil {
+		return nil, err
+	}
+
+	addr, err := manual.HostAddress(hostname)
+	if err != nil {
+		return nil, errors.Annotatef(err, "failed to compute public address for %q", hostname)
+	}
+
+	provisioned, err := checkProvisioned(hostname, cli)
+	if err != nil {
+		return nil, errors.Annotatef(err, "cannot decide if the machine is provisioned")
+	}
+	if provisioned {
+		return nil, manual.ErrProvisioned
+	}
+
+	hc, series, err := DetectSeriesAndHardwareCharacteristics(hostname, cli)
+	if err != nil {
+		err = fmt.Errorf("error detecting windows characteristics: %v", err)
+		return nil, err
+	}
+	// There will never be a corresponding "instance" that any provider
+	// knows about. This is fine, and works well with the provisioner
+	// task. The provisioner task will happily remove any and all dead
+	// machines from state, but will ignore the associated instance ID
+	// if it isn't one that the environment provider knows about.
+
+	instanceId := instance.Id(manual.ManualInstancePrefix + hostname)
+	nonce := fmt.Sprintf("%s:%s", instanceId, uuid)
+	machineParams := &params.AddMachineParams{
+		Series:                  series,
+		HardwareCharacteristics: hc,
+		InstanceId:              instanceId,
+		Nonce:                   nonce,
+		Addrs:                   params.FromNetworkAddresses(addr),
+		Jobs:                    []multiwatcher.MachineJob{multiwatcher.JobHostUnits},
+	}
+	return machineParams, nil
+}
+
+// checkProvisioned checks if the machine is already provisioned or not.
+// if it's already provisioned it will return true
+func checkProvisioned(host string, cli manual.WinrmClientAPI) (bool, error) {
+	logger.Tracef("Checking if %s windows machine is already provisioned", host)
+	var stdout, stderr bytes.Buffer
+
+	// run the command trough winrm
+	// this script detects if the jujud process, service is up and running
+	script, err := shell.NewPSEncodedCommand(detectJujudProcess)
+	if err != nil {
+		return false, errors.Trace(err)
+	}
+
+	// send the script to the windows machine
+	if err = cli.Run(script, &stdout, &stderr); err != nil {
+		return false, errors.Trace(err)
+	}
+
+	provisioned := strings.Contains(stdout.String(), "Yes")
+	if stderr.Len() != 0 {
+		err = errors.Annotatef(err, "%v (%v)", err, strings.TrimSpace(stderr.String()))
```

Comment:
Annotatef already includes the original error in the message. Please change this to:

`err = errors.Annotate(err, strings.TrimSpace(stderr.String()))`